lesson4/README.md

```diff
@@ -1,9 +1,93 @@
 # Lesson 4 - Data Storage App Life Cycle
 
 ##  Data Storage
+Everytime you close an app and reopen it, a new session of app is created. This means that any data that wasn't saved cannot be retrieved again. This is where persistent storage comes in. Android provides you with 5-6 options to choose from. While choosing an option, you should consider complexity of data being stored, privacy (if it can be read by the app only, the users or other apps on the device), and location (if it is local or on the cloud). We'll be mainly discussing Shared Preferences and SQLite databases. But here are many of the options available to Android:
+
+1. [Shared Preferences](https://developer.android.com/guide/topics/data/data-storage.html#pref):
+
+   Useful for writing small amount of data as key-value pairs.
+2. [Firebase](https://www.firebase.com/docs/android/quickstart.html):
+
+   A cool way to store data, run tests and analytics, authenticate, and monetize. The quickstart page linked here gives you a very good idea about how to use Firebase.
+3. [Internal Storage](https://developer.android.com/guide/topics/data/data-storage.html#filesInternal):
+
+   This memory chunk is private by default and is only available to the app. This data gets deleted when the app is uninstalled from the Android device. You can use methods like `openFileInput()`, `read()`, and `close()` on a [FileOutputStream](https://developer.android.com/reference/java/io/FileOutputStream.html)
+4. [External Storage](https://developer.android.com/guide/topics/data/data-storage.html#filesExternal):
+
+   The data stored here is world-readable but the data itself can't be relied upon as the user can unmount the storage anytime they want. Be sure to handle in the code the case where the data is unavailable.
+   ```java
+    /* Checks if external storage is available for read and write */
+    public boolean isExternalStorageWritable() {
+        String state = Environment.getExternalStorageState();
+        if (Environment.MEDIA_MOUNTED.equals(state)) {
+            return true;
+        }
+        return false;
+    }
+    
+    /* Checks if external storage is available to at least read */
+    public boolean isExternalStorageReadable() {
+        String state = Environment.getExternalStorageState();
+        if (Environment.MEDIA_MOUNTED.equals(state) ||
+            Environment.MEDIA_MOUNTED_READ_ONLY.equals(state)) {
+            return true;
+        }
+        return false;
+    }
+   ```
+5. [Cache](https://developer.android.com/reference/android/content/Context.html#getCacheDir()):
+
+   If you have data which qualifies this: "it would be nice to save it somewhere instead of making users type but it is no big deal if the data gets deleted", use cache. Remember that you can't store large amounts or data (>1MB) without causing problems. Also, be responsible about how you use the cache memory. Clean the space you occupied after you are done.
```

Comment:
is the cache basically for simple stuff like form/login info? What's the primary use case?